### DISCUSSION TITLE
Dates not parsed from useLoaderData ?

### DISCUSSION THREAD

**Discussion by QuentinWidlocher** [remix-run/remix#634] [Q&A] [answered]
Hi everyone,

I returned an object with a date property from my `loader` function and got back a string inside my object instead of a Date.
Is it a normal behavior ? 

I can map my object property from a string to a date in the front-end of course but it's a bit cumbersome and Remix already makes my life so easier, so I'm surprised !

**Comment by sergiodxa** [comment, marked as answer]
Any data you return from a loader, or if you use the `json` helper, will be sent to `JSON.stringify`, so Date objects will be converted to strings, then when Remix reads the data it will do `JSON.parse` of that string, but `JSON.parse` is not going to revive the date string as a Date object so you will get a string from `useLoaderData`.

**Reply by sergiodxa** [reply]
I opened an issue requesting to use superjson so Date works https://github.com/remix-run/remix/issues/656

**Reply by QuentinWidlocher** [reply]
Thanks for opening this issue, this could be very beneficial to Remix. Having a seamless relationship between the front and the back is so awesome

**Reply by iamstiil** [reply]
When using `JSON.stringify`/`JSON.parse` you can pass in a replacer/reviver function in which you can serialize/deserialize fields how see fit. I think this should be enough to cover all needs.

**Reply by sergiodxa** [reply]
Remiz doesn’t let you control that, so you can pass a custom reviver function, and when using response.json you can’t pass such function.

**Reply by jjhiggz** [reply]
I think it would be really nice to at least have some kind of option in `useLoaderData` that could allow you to serialize dates so that way you get a better typescript experience. 

Right now my problem is that, rightfully so, I want to get  a `Restaurant` (from a prisma model) type out of loader data, because I have other frontend stuff that is built to take in a `Restaurant`, but I am instead getting something like this

```ts
// this will no longer be recognized as a restaurant
{
    name: string
    description: string
    createdAt: string // not date
} 
```

What would be cool is if there was something I could pass into `useLoaderData` like so:

```ts
const data = useLoaderData({
      serializeDates: true,
})
```

Or maybe if you want to specify fields

```ts
const data = useLoaderData({
    serializeDates: {
          createdAt: true
          updatedAt: true
    }
})
```

now the type of `data.restaurant` could be 

```ts
{
    name: string
    description: string
    createdAt: Date
} 
```
which I can now pass into any function expecting a restaurant

**Reply by QuentinWidlocher** [reply]
I use [superjson](https://github.com/blitz-js/superjson) for this exact purpose and it works very well.
A built-in solution would be nice though, I agree.

**Reply by sergiodxa** [reply]
In Remix Utils there's a custom `json`, `useLoaderData` and `useActionData` which let you customize how you serialize and parse values, this can let you revive dates from the string.

**Reply by kiliman** [reply]
@QuentinWidlocher I added superjson to the new type inference released in v1.6.5. So you get full type fidelity from loader/action to route component.

https://github.com/kiliman/remix-superjson

**Reply by QuentinWidlocher** [reply]
Wow that's great thanks !
I know it was a bit of a debate but I think it's a perfect choice for Remix

**Reply by kiliman** [reply]
I'm going to see if I can come up with a paired down version of superjson, that works for most cases. Right now superjson adds 9KB to a production bundle.

**Reply by candidia** [reply]
@sergiodxa I can't find an export of the custom useLoaderData you mentioned. Did you remove this, and what is your preferred solution if so?

**Reply by mjjabarullah** [reply]
![Screenshot 2024-11-03 130722](https://github.com/user-attachments/assets/5a87d609-e69a-4147-a13b-3f2653fea55c)

I just tried remix-superjson. But it has type error with superjson. Does this support latest version of remix?
